File: daily_problems/2024/11/1122/personal_submission/cf215e_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/fenwicktree"
// #include "atcoder/lazysegtree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

long long f(long long x) {
    int k = (x ? 64 - __builtin_clzll(x) : 0);
    vector<long long> ans(k + 1, 0);

    for (int l = 1; l < k; l ++) {
        for (int divisor = 1; divisor < l; divisor ++) {
            if (l % divisor == 0) {
                ans[l / divisor] += 1ll << (divisor - 1);
            }
        }
    }

    for (int divisor = 1; divisor < k; divisor ++) {
        if (k % divisor == 0) {
            long long msk = 0;
            for (int j = 0; j < k; j += divisor) {
                msk |= 1ll << j;
            }
            ans[k / divisor] += max(0ll, x / msk - (1ll << (divisor - 1)) + 1);
        }
    }

    long long res = 0;
    for (int i = k; i > 1; i --) {
        for (int j = i * 2; j <= k; j += i)
            ans[i] -= ans[j];
        res += ans[i];
    }

    return res;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    long long l, r;
    cin >> l >> r;
    cout << f(r) - f(l - 1);

    return 0;
}

File: daily_problems/2024/11/1122/personal_submission/cf215e_yawn_sean.py
# Submission link: https://codeforces.com/contest/215/submission/292695199
def main():
    def f(x):
        k = x.bit_length()
        ans = [0] * (k + 1)
        for l in range(1, k):
            for divisor in range(1, l):
                if l % divisor == 0:
                    ans[l // divisor] += 1 << divisor - 1
        
        for divisor in range(1, k):
            if k % divisor == 0:
                msk = 0
                for j in range(0, k, divisor):
                    msk |= 1 << j
                ans[k // divisor] += fmax(0, x // msk - (1 << divisor - 1) + 1)
        
        res = 0
        for i in range(k, 1, -1):
            for j in range(i * 2, k + 1, i):
                ans[i] -= ans[j]
            res += ans[i]
        
        return res

    l, r = MII()
    print(f(r) - f(l - 1))